FEPointEvaluation: Allow evaluating more points together

kronbichler opened the pull request:
I noticed in a benchmark (written by @fdrmrc) that `MGTwoLevelTransferNonNested` spends too much time in evaluating the polynomial basis. More precisely, there is quite some overhead (both in terms of instructions and latency chains) in the formula https://github.com/dealii/dealii/blob/81ec8648500a910743e28cef437c843d84a469e6/include/deal.II/base/polynomial.h#L924-L934 With the present PR, I enable the code to process two points together, which essentially puts all overhead and latency into half. It would not make sense to use more than 2 points in 3D, because the current algorithm requires 6 registers for the point coordinates, 6 registers for the accumulated products, and at least 1 register for the `lagrange_support_points[i]` variable, and there are still instruction sets with only 16 FP registers. (We could use 3 points in 2D, or in the future when 32 registers are widespread enough also in 3D.)

Also note that I only do this when evaluating values only (because the derivatives need a second set of `dim * n_points` registers for the derivative variable; they already have twice the arithmetic work so less critical). This PR will be most effective when combined with inlining of the relevant function, which I will do in a separate PR.

kronbichler commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -2074,12 +2073,26 @@ FEPointEvaluationBase<n_components_, dim, spacedim, Number>::do_reinit()
           }
         else
           {
-            shapes.resize_fast(n_q_batches * n_shapes);
```
This line (and a similar one for the faces above) is only tangentially related to the combination of points. However, it is still crucial: Even though `resize_fast` does not do much, https://github.com/dealii/dealii/blob/81ec8648500a910743e28cef437c843d84a469e6/include/deal.II/base/aligned_vector.h#L1298-L1304 when the size is already correct, we should nonetheless only do the operation once outside the loop over `n_q_batches`, not in every loop iteration.

kronbichler commented:
FYI @bergbauer

gassmoeller commented on include/deal.II/matrix_free/fe_point_evaluation.h:
```diff
@@ -2074,12 +2073,26 @@ FEPointEvaluationBase<n_components_, dim, spacedim, Number>::do_reinit()
+              }
+            else if (qb + 1 < n_q_batches)
+              {
+                internal::compute_values_of_array_in_pairs(
```
Maybe leave a comment here that explains why you compute pairs?

gassmoeller commented on include/deal.II/matrix_free/tensor_product_point_kernels.h:
```diff
@@ -75,6 +75,32 @@ namespace internal
+
+
   /**
    * Specialization of above function for dim = 0. Should not be called.
```
This comment is now confusing, because you introduced the additional function. Maybe expand to which of the above functions this one is a specialization of.

gassmoeller commented on include/deal.II/matrix_free/tensor_product_point_kernels.h:
```diff
@@ -75,6 +75,32 @@ namespace internal
+  compute_values_of_array_in_pairs(
+    dealii::ndarray<Number, 2, dim>                    *shapes,
+    const std::vector<Polynomials::Polynomial<double>> &poly,
+    const Point<dim, Number>                           *p)
```
I am not sure if this was discussed in previous reviews, but are raw pointers here necessary? In this case could the function just take a const reference to `p1` and another one to `p2`? That way there would be no danger of someone handing over a pointer to an array that doesnt have 2 entries.

kronbichler replied:
No, the raw pointers should not be necessary here. I checked with my compiler and what I see is two more integer additions to pass over the second reference (in an additional register that is part of the calling convention on my x86-64 system), but I think that is acceptable for the increase in readability.

gassmoeller commented on include/deal.II/matrix_free/tensor_product_point_kernels.h:
```diff
@@ -75,6 +75,32 @@ namespace internal
+    const std::vector<Polynomials::Polynomial<double>> &poly,
+    const Point<dim, Number>                           *p)
+  {
+    const int n_shapes = poly.size();
```
Just out of curiosity, why is this `int` and not `unsigned int` or `size_type`? Is this just copy-paste from the existing function, or is there some reason behind it?

gassmoeller reviewed (commented):
I am always happy about performance improvements :-)
Can you add some documentation and consider the questions I marked up?

kronbichler commented:
I addressed the review comments, thank you both for the helpful input. I took the question as an opportunity to add some comments in the code, there might be others having the same or similar questions.

gassmoeller reviewed (approved):
Looks good to me, thanks for the explanations. Out of curiosity, do you have some numbers on how much this saves compared to the other function?

Ready to go from my side.

kronbichler commented:
On my machine, the restriction and prolongation got around 10% faster, but that was in combination with #16911. (And combined with #16896 the whole multigrid solver was around 5-8% faster.) Isolated to this function, I see an advantage of around 5% for a 3D multigrid solution in the transfer operations within the multigrid solver (2.12e-2 seconds versus 1.98e-2 seconds for restriction, 1.60e-2 s vs 1.46e-2 s for prolongation).
